File: packages/mirascript/src/vm/operations/compound.ts
import { VmError } from '../../helpers/error.js';
import { hasOwnEnumerable, NotNumber, isFinite, getRecord, keysLength } from '../../helpers/utils.js';
import { toNumber } from '../../helpers/convert/index.js';
import { display } from '../../helpers/serialize.js';
import { isVmArray, isVmRecord, isVmExtern, isVmWrapper } from '../../helpers/types.js';
import type { VmAny, VmValue } from '../types/index.js';
import { $AssertInit } from './common.js';
import { $ToString } from './convert.js';
import { $El } from './helpers.js';

const { trunc } = Math;
const { at } = Array.prototype;

/** 获取值的长度 */
export const $Length = (value: VmAny): number => {
    $AssertInit(value);
    if (isVmArray(value)) return value.length;
    if (isVmRecord(value)) return keysLength(value);
    if (isVmWrapper(value)) return value.keys().length;

    throw new VmError(`Value has no length: ${display(value)}`, 0);
};

/** 检查是否拥有字段 */
export const $Has = (obj: VmAny, key: VmAny): boolean => {
    $AssertInit(obj);
    const pk = $ToString(key);
    if (obj == null || typeof obj != 'object') return false;
    if (isVmWrapper(obj)) return obj.has(pk);
    return hasOwnEnumerable(obj, pk);
};

/** 获取字段 */
export const $Get = (obj: VmAny, key: VmAny): VmValue => {
    $AssertInit(obj);
    if (isVmArray(obj)) {
        $AssertInit(key);
        const index = toNumber(key, NotNumber);
        if (!isFinite(index)) return null;
        return $El((at.call(obj, trunc(index)) ?? null) as VmAny);
    }
    const pk = $ToString(key);
    if (obj == null || typeof obj != 'object') return null;
    if (isVmWrapper(obj)) {
        if (isFinite(key) && isVmExtern(obj) && obj.isArrayLike()) {
            let index = trunc(key as number);
            const { length } = obj.value;
            if (index < 0) index += length;
            if (index >= 0 && index < length) {
                return obj.get(String(index)) ?? null;
            }
        }
        return obj.get(pk) ?? null;
    }
    return $El(getRecord(obj, pk) ?? null);
};
/** 设置字段 */
export const $Set = (obj: VmAny, key: VmAny, value: VmAny): void => {
    $AssertInit(obj);
    $AssertInit(value);
    const pk = $ToString(key);
    if (!isVmExtern(obj)) throw new VmError(`Expected extern, got ${display(obj)}`, undefined);
    obj.set(pk, value);
};

File: crates/python/mirascript/_vm/operations/compound.py
from __future__ import annotations
import math

from ..._helpers.convert import to_number
from ..._helpers.types import (
    is_vm_array,
    is_vm_extern,
    is_vm_primitive,
    is_vm_record,
    is_vm_wrapper,
)
from ..types import VmAny, VmValue
from ..error import VmError
from .common import AssertInit
from .convert import ToString
from .type_check import Type
from .helpers import Element


def Length(a: VmAny) -> float:
    AssertInit(a)
    if isinstance(a, (list, dict)):
        return float(len(a))
    raise TypeError(f"`Expected array, record or extern, got {Type(a)}")


def Has(obj: VmAny, key: VmAny) -> bool:
    pk = ToString(key)
    if is_vm_wrapper(obj):
        return obj.has(pk)
    if is_vm_record(obj):
        return pk in obj
    if is_vm_array(obj):
        try:
            idx = to_number(key)
            if not idx.is_integer():
                return False
            return 0 <= idx < len(obj)
        except Exception:
            return False
    if is_vm_primitive(obj):
        return False
    AssertInit(obj)
    return False


def Get(obj: VmAny, key: VmAny) -> VmValue:
    if is_vm_array(obj):
        AssertInit(key)
        index = to_number(key, math.nan)
        if not math.isfinite(index):
            return None
        try:
            idx = math.trunc(index)
            return Element(obj[idx])
        except IndexError:
            return None
    if is_vm_primitive(obj):
        return None
    pk = ToString(key)
    if is_vm_wrapper(obj):
        return obj.get(pk)
    if is_vm_record(obj):
        return Element(obj.get(pk, None))
    AssertInit(obj)
    return None


def Set(obj: VmAny, key: VmAny, val: VmAny) -> None:
    pk = ToString(key)
    if not is_vm_extern(obj):
        AssertInit(obj)
        raise VmError(f"`Expected extern object, got {Type(obj)}", None)

    AssertInit(val)
    obj.set(pk, val)
